From da36e24d473ddf316468933a1116570dca1f7844 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:08:50 -0400
Subject: [usb] usb_wwan: clean up delayed-urb submission

Message-id: <1410887463-20674-459-git-send-email-dzickus@redhat.com>
Patchwork-id: 95369
O-Subject: [RHEL7 PATCH 458/591] USB: usb_wwan: clean up delayed-urb submission
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit 3362c91c7841f2e4ab5b1f018a5da0ecbc924894
Author: Johan Hovold <jhovold@gmail.com>
Date:   Mon May 26 19:23:27 2014 +0200

    USB: usb_wwan: clean up delayed-urb submission

    Clean up and rename delay-urb submission function using a more
    descriptive name.

    Also add comment on locking assumptions.

    Signed-off-by: Johan Hovold <jhovold@gmail.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/serial/usb_wwan.c b/drivers/usb/serial/usb_wwan.c
index b83aa60..45bc11b 100644
--- a/drivers/usb/serial/usb_wwan.c
+++ b/drivers/usb/serial/usb_wwan.c
@@ -602,7 +602,8 @@ int usb_wwan_suspend(struct usb_serial *serial, pm_message_t message)
 }
 EXPORT_SYMBOL(usb_wwan_suspend);
 
-static int play_delayed(struct usb_serial_port *port)
+/* Caller must hold susp_lock. */
+static int usb_wwan_submit_delayed_urbs(struct usb_serial_port *port)
 {
  struct usb_serial *serial = port->serial;
  struct usb_wwan_intf_private *data = usb_get_serial_data(serial);
@@ -613,11 +614,14 @@ static int play_delayed(struct usb_serial_port *port)
 
  portdata = usb_get_serial_port_data(port);
 
- while ((urb = usb_get_from_anchor(&portdata->delayed))) {
+ for (;;) {
+  urb = usb_get_from_anchor(&portdata->delayed);
+  if (!urb)
+   break;
+
   err = usb_submit_urb(urb, GFP_ATOMIC);
   if (err) {
-   dev_err(&port->dev,
-     "%s: submit write urb failed: %d\n",
+   dev_err(&port->dev, "%s: submit urb failed: %d\n",
      __func__, err);
    err_count++;
    unbusy_queued_urb(urb, portdata);
@@ -664,7 +668,7 @@ int usb_wwan_resume(struct usb_serial *serial)
    }
   }
 
-  err = play_delayed(port);
+  err = usb_wwan_submit_delayed_urbs(port);
   if (err)
    err_count++;
 
-- 
1.7.1